crates/packet_inspector/src/main.rs

```diff
@@ -23,107 +34,79 @@ use valence_protocol::packet::c2s::status::{QueryPingC2s, QueryRequestC2s};
-}
-
-struct State {
-    cli: Arc<Cli>,
-    enc: PacketEncoder,
-    dec: PacketDecoder,
-    read: OwnedReadHalf,
-    write: OwnedWriteHalf,
-    buf: String,
-    style: owo_colors::Style,
-}
-
-impl State {
-    pub async fn rw_packet<'a, P>(&'a mut self) -> anyhow::Result<P>
-    where
-        P: Packet<'a>,
-    {
-        while !self.dec.has_next_packet()? {
-            self.dec.reserve(4096);
-            let mut buf = self.dec.take_capacity();
-
-            if self.read.read_buf(&mut buf).await? == 0 {
-                return Err(io::Error::from(ErrorKind::UnexpectedEof).into());
-            }
-
-            self.dec.queue_bytes(buf);
-        }
-
-        let pkt: P = self.dec.try_next_packet()?.unwrap();
-
-        self.enc.append_packet(&pkt)?;
-
-        let bytes = self.enc.take();
-        self.write.write_all(&bytes).await?;
-
-        self.buf.clear();
-        write!(&mut self.buf, "{pkt:?}")?;
 
-        let packet_name = self
-            .buf
-            .split_once(|ch: char| !ch.is_ascii_alphabetic())
-            .map(|(fst, _)| fst)
-            .unwrap_or(&self.buf);
+    /// Disable the GUI. Logging to stdout.
+    #[clap(long)]
+    nogui: bool,
 
-        if let Some(r) = &self.cli.include_regex {
-            if !r.is_match(packet_name) {
-                return Ok(pkt);
-            }
-        }
-
-        if let Some(r) = &self.cli.exclude_regex {
-            if r.is_match(packet_name) {
-                return Ok(pkt);
-            }
-        }
-
-        println!("{}", self.buf.style(self.style));
+    /// Only show packets that match the filter.
+    #[clap(short, long)]
+    include_filter: Option<Regex>,
 
-        Ok(pkt)
-    }
+    /// Hide packets that match the filter. Note: Only in effect if nogui is
+    /// set.
+    #[clap(short, long)]
+    exclude_filter: Option<Regex>,
 }
 
 #[tokio::main]
-async fn main() -> Result<(), Box<dyn Error>> {
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
     tracing_subscriber::fmt()
         .with_max_level(LevelFilter::DEBUG)
         .init();
 
     let cli = Arc::new(Cli::parse());
 
+    match cli.nogui {
+        true => start_cli(cli).await?,
+        false => start_gui(cli)?,
+    };
+
+    Ok(())
+}
+
+async fn start_cli(cli: Arc<Cli>) -> Result<(), Box<dyn std::error::Error>> {
+    let context = Arc::new(Context::new(ContextMode::Cli(context::Logger {
+        include_filter: cli.include_filter.clone(),
+        exclude_filter: cli.exclude_filter.clone(),
+    })));
+
     let sema = Arc::new(Semaphore::new(cli.max_connections.unwrap_or(100_000)));
 
-    eprintln!("Waiting for connections on {}", cli.client_addr);
-    let listen = TcpListener::bind(cli.client_addr).await?;
+    let client_addr = match cli.client_addr {
+        Some(addr) => addr,
+        None => panic!("client_addr must be set"),
+    };
+
+    let server_addr = match cli.server_addr {
+        Some(addr) => addr,
+        None => panic!("server_addr must be set"),
```

Comment:
Technically, if server and client addr are None at this point, it's a bug, since Clap enforces a value with
```
#[arg(required_if_eq("nogui", "true"))]
```

But yeah, returning an error should not be that different in that both would do an exit.